### PR TITLE
Add more infomation in FileSourceScanExec log when timezone is not UTC

### LINKED ISSUE #6213: Reading ORC file with timestamp when timezone not UTC doesn't give much information
**Is your feature request related to a problem? Please describe.**
When reading a ORC file that has a TimestampType but your timezone isn't set to UTC we just print:

`!Exec <FileSourceScanExec> cannot run on GPU because unsupported data types in output: TimestampType [_c9#9]; unsupported data types TimestampType [_c9] in read for ORC`

It would be nice to print that this is because your timezone isn't in UTC.

### PR DESCRIPTION
**Pull request by thirtiseven**
When reading a ORC file that has a TimestampType but your timezone isn't set to UTC we just print:

`!Exec <FileSourceScanExec> cannot run on GPU because unsupported data types in output: TimestampType [_c9#9]; unsupported data types TimestampType [_c9] in read for ORC`

It didn't give enough information.

This PR add a function in FileFormatChecks to check if the file contains TimestampType and the timezone isn't set to UTC. If so, we will add an additional warning message in FileSourceScanExec log. 

`!Exec <FileSourceScanExec> cannot run on GPU because unsupported data types TimestampType [_c0] in read for ORC; your timezone isn't in UTC (JVM: GMT-08:00, session: GMT-08:00). Set both of the timezones to UTC to enable TimestampType support; unsupported data types in output: TimestampType [_c0#0]`

This change also works for parquet files.

Closes #6213 

Signed-off-by: thirtiseven <ntlihy@gmail.com>

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 1]
```diff
@@ -791,6 +792,22 @@ abstract class TypeChecks[RET] {
...
+
+  /** Add more infomation in log when timezone not UTC */
+  protected def tagTimestampUtcCheck(
+    meta: RapidsMeta[_, _, _],
+    fields: Seq[StructField]
+    ): Unit = {
+    val schemaHasTimestamps = fields.exists { field =>
+      TrampolineUtil.dataTypeExistsRecursively(field.dataType, _.isInstanceOf[TimestampType])
+    }
+    if (schemaHasTimestamps && !(TypeChecks.areTimestampsSupported(ZoneId.systemDefault()) &&
```
Comment (res-life) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:804]: Duplicated, can we extract a `util` method?

https://github.com/NVIDIA/spark-rapids/blob/branch-22.10/sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala#L423toL429

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 2]
```diff
@@ -905,6 +922,7 @@ class FileFormatChecks private (
       schema: StructType,
       fileType: FileFormatType,
       op: FileFormatOp): Unit = {
+    tagTimestampUtcCheck(meta, schema.fields)
     tagUnsupportedTypes(meta, sig, schema.fields,
```
Comment (res-life) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:926]: Checked twice: `tagTimestampUtcCheck`, `tagUnsupportedTypes`
Can we update `tagUnsupportedTypes` to return the error string?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 3]
```diff
@@ -776,6 +776,23 @@ abstract class TypeChecks[RET] {
     }.mkString(", ")
   }
 
+  private def tagCustomMessage(
+    groupedByType: Map[DataType, Set[String]],
+    meta: RapidsMeta[_, _, _]
+    ): Map[DataType, Set[String]] = {
+    groupedByType.filterKeys {
```
Comment (res-life) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:783]: Seems can not check nested types like `Array(timestamp)`

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 4]
```diff
@@ -776,6 +776,23 @@ abstract class TypeChecks[RET] {
...
+      case TimestampType if !TypeChecks.areTimestampsSupported(ZoneId.systemDefault()) ||
```
Comment (res-life) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:784]: Nit, Extract a method for the following check:
```
          TypeChecks.areTimestampsSupported(ZoneId.systemDefault()) &&
          TypeChecks.areTimestampsSupported(SQLConf.get.sessionLocalTimeZone)
```

Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala]: done

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala (hunk 3 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala]: done

**Comment by thirtiseven** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala [hunk 5]
```diff
@@ -776,6 +774,35 @@ abstract class TypeChecks[RET] {
...
+    meta: RapidsMeta[_, _, _]
+    ): Map[DataType, Set[String]] = {
+    def checkTimestampType(dataType: DataType): Boolean = dataType match {
+        case TimestampType if !TypeChecks.areTimestampsSupported() => {
+          meta.willNotWorkOnGpu(s"your timezone isn't in UTC (JVM:" +
+            s" ${ZoneId.systemDefault()}, session: ${SQLConf.get.sessionLocalTimeZone})." +
+            s" Set both of the timezones to UTC to enable TimestampType support")
+          false
+        }
+        case ArrayType(elementType, _) =>
```
Comment (res-life) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:792]: If one column is `array(timestamp)` , `ArrayType` is unsupported, `TimestampType` is supported and time zone is not UTC, this method returns false, `tagTimestampTypeAndReturnOthers` will remove this `array(timestamp)` type.
Actually, we can't distinguish the unsupported reasons for nested types.
Let's just add one extra error msg of UTC and remain the original error msg.

Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala:792]: Nice catch, fixed.

**Comment by res-life** [comment]
LGTM

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

**Review by res-life — APPROVED** [review]

**Pull request merged**